modules/gapi/src/backends/ie/giebackend.cpp

```diff
@@ -480,6 +479,166 @@ inline IE::Blob::Ptr extractBlob(IECallContext& ctx, std::size_t i) {
+}
+
+void cv::gimpl::ie::RequestPool::wait_and_shutdown() {
+    {
+        std::lock_guard<std::mutex> lk{m_task_mutex};
+        // Check if it's already shutdown.
+        if (m_is_shutdown) {
+            return;
+        }
+        m_is_shutdown = true;
+    }
+
+    m_task_cv.notify_one();
+
+    for (auto&& worker : m_workers) {
+        worker.join();
+    }
+}
+
+cv::gimpl::ie::RequestPool::~RequestPool() {
+    wait_and_shutdown();
+}
+
+void cv::gimpl::ie::RequestPool::callback(cv::gimpl::ie::RequestPool::Task task,
+                                          cv::gimpl::ie::RequestPool::CompiledPtr compiled,
+                                          size_t id) {
+    task.callback(*compiled);
+    {
+        std::lock_guard<std::mutex> idle_lock(m_idle_mutex);
+        m_idle_ids.push(id);
+    }
+    m_idle_cv.notify_one();
+}
+
+void cv::gimpl::ie::RequestPool::worker() {
+    // The worker algorigthm:
+    // In inifinity loop:
+    // 1. Worker wait for not empty task queue or wait_and_shutdown is called.
+    // 2. If worker woke up and queue isn't empty then.
+    //     2.1 Get a task from queue and release lock.
+    //     2.2 Wait for idle IECompiled.
+    //     2.3 Get the id of the idle IECompiled
+    //     2.4 Drop this id from idle queue, which means that IECompiled isn't idle anymore.
+    //     2.5 Get IECompiled corresponing to id and release lock.
+    //     2.6 Run task and pass IECompiled id to the callback which returns it back to the queue.
+    //         Then go to the next iteration.
+    // 3. In this execution moment there is no tasks in queue and shutdown is called.
+    //    But some compiled networks probably working at that moment, let's check it.
+    // 4. If all compiled networks are idle, need to double check a task queue,
+    //    because the tasks can submit the new ones in their callbacks, it that case go to (1).
+    // 5. Finally if all compiled networks are idle and task queue is empty, finish execution.
+
+    while (true) {
+        // (1) Wait for a task or shutdown.
+        {
+            std::unique_lock<std::mutex> task_lk{m_task_mutex};
+            m_task_cv.wait(task_lk, [this]() { return !m_tasks.empty() || m_is_shutdown; });
+
+            if (!m_tasks.empty()) {
+                // (2.1) Get task from queue.
+                auto task = m_tasks.front();
+                m_tasks.pop();
+                task_lk.unlock();
+
+                // (2.2) Wait for idle IECompiled.
+                std::unique_lock<std::mutex> idle_lock(m_idle_mutex);
+                m_idle_cv.wait(idle_lock, [this](){ return !m_idle_ids.empty(); });
+
+                // (2.3) Get the id of the first idle IECompiled.
+                size_t id = m_idle_ids.front();
+                // (2.4) Drop from queue.
+                m_idle_ids.pop();
+                // (2.5) Get the first idle IECompiled by id.
+                auto& compiled = m_compiled_nets[id];
+                idle_lock.unlock();
```

Comment:
I'd use scopes instead. At least they work automatically, e.g. when an exception occur.

Reply:
Not relevant